Regen network/multistore upgrades

anilCSE opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: https://github.com/regen-network/cosmos-sdk/issues/7

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [x] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [x] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

ethanfrey commented on x/upgrade/internal/types/storeloader.go:
```diff
@@ -0,0 +1,129 @@
+		}
+
+		// Check if the current commit version and upgrade height matches
+		if (len(upgrades.StoreUpgrades.Renamed) > 0 || len(upgrades.StoreUpgrades.Deleted) > 0) &&
```
This looks like similar logic (but cleaned up) from the original upgrade handler.

Big issue here, is that we now never write the store upgrades from the pre-upgrade binary (design was that they would  be in the proposal), but use them from the post-upgrade binary (written in code). I understand the reason behind this change, however, here we check value from file.

I would actually remove the `StoreUpgrades` field form UpgradeInfo as it is confusing and not used elsewhere in the code (I think). And you should be checking eg. `len(storeUpgrades.Renamed)` here. I see you do use the proper `storeUpgrades` as an argument to `ms.LoadLatestVersionAndUpgrade`, so it seems that you just missed to upgrade the if statement.

sahith-narahari replied:
Yeah,sorry missed this. Changed in next commit

ethanfrey commented on x/upgrade/internal/types/storeloader.go:
```diff
@@ -0,0 +1,129 @@
+//
+// This is useful for in place migrations when a store key is renamed between
+// two versions of the software.
+func UpgradeableStoreLoader(upgradeInfoPath string) baseapp.StoreLoader {
```
This is an implementation of the older technique, right?
The code looks correct to me, but I am a bit confused when and where we would use one or the other.

sahith-narahari replied:
This was to address store changes saved to disk, but as that is deprecated I'll remove the one which takes path

ethanfrey commented on x/upgrade/internal/types/storeloader_test.go:
```diff
@@ -0,0 +1,147 @@
+
+	upgradeInfoFilePath := filepath.Join(homeDir, "upgrade-info.json")
+
+	data := []byte(`{"height": 0, "store_upgrades": {"renamed":[{"old_key": "bnk", "new_key": "bank"}]}}`)
```
I assume to use a non-zero height? We want to use the same height on BeginBlock, right? They are supposed to check equality.

If this is eg. `height=2`, then we mount as below and show that *no migrations happen* (even with the migration handler installed). Then commit eg. 2 blocks, so the iavl tree is now stored at height 2. Then mount the store again (with committed height =2) and validate the stores were renamed properly.

A bit more complicated, but important to ensure the checks for "only at height" work out properly

sahith-narahari replied:
Yes, will clean this up

jackzampolin commented on simapp/app.go:
```diff
@@ -196,7 +196,7 @@ func NewSimApp(
 		app.subspaces[crisis.ModuleName], invCheckPeriod, app.SupplyKeeper, auth.FeeCollectorName,
 	)
-	app.UpgradeKeeper = upgrade.NewKeeper(skipUpgradeHeights, keys[upgrade.StoreKey], app.cdc)
+	app.UpgradeKeeper = upgrade.NewKeeper(skipUpgradeHeights, keys[upgrade.StoreKey], app.cdc, DefaultNodeHome)
```
@ethanfrey that is up to the specific application. You can see how that is handled in gaia

ethanfrey replied:
Actually the whole home processing stuff is burried in the client libs in cosmos-sdk not in the app. The default is set by app, but how we override with `--home` or `--home-client` is standardized

jackzampolin commented on x/upgrade/internal/types/storeloader.go:
```diff
@@ -0,0 +1,129 @@
+
+// StoreLoaderWithUpgrade is used to prepare baseapp with a fixed StoreLoader
+// pattern. This is useful in test cases, or with custom upgrade loading logic.
+func StoreLoaderWithUpgrade(storeUpgrades *storetypes.StoreUpgrades) baseapp.StoreLoader {
```
keeping viper calls out of the modules and pushing them up into the cobra cli code is preferred.

aaronc commented:
> So, when preparing for an upgrade, we would have two lines:
> 
> ```go
> # prepare the next update upgrade handler
> upgradeKeeper.SetHandler("next-update", migrationFunc)
> # only if we want to do upgrades in next update 
> storeLoader := upgrade.StoreLoaderWithUpgrade(storeUpgradesForNextUpdate)
> app.SetStoreLoader(storeLoader)
> ```

Since this is a pretty infrequent operation, I think something that puts the burden in `app.go` is okay. Basically somewhere, maybe even just in `app.go` there needs to be some logic that:

1. reads upgrade height and upgrade name from `upgrade-info.json`
2. checks if there are are store upgrades for this upgrade name
3. checks if the current store version is equal to the upgrade height in `upgrade-info.json`
4. if so, apply the store upgrades in initialization

So this could be as simple as something like:

```go
upgradeName, upgradeHeight := upgradeKeeper.ReadUpgradeInfoFromDisk()
if upgradeName == "myUpgrade" {
  // configure store loader that checks if version == upgradeHeight and applies store upgrades
  app.SetStoreLoader(UpgradeStoreLoader(upgradeHeight, storeUpgrades)
}
```

This could also be more self-contained if we added a store upgrades parameter to `SetHandler`:
```go
upgradeKeeper.SetHandler("next-update", migrationFunc, storeUpgrades)
app.SetStoreLoader(upgradeKeeper.UpgradeStoreLoader())
```

Happy to get on a quick call to talk this through if that would help.

anilCSE commented:
> This could also be more self-contained if we added a store upgrades parameter to `SetHandler`:
> 
> ```go
> upgradeKeeper.SetHandler("next-update", migrationFunc, storeUpgrades)
> app.SetStoreLoader(upgradeKeeper.UpgradeStoreLoader())
> ```

I agree with @ethanfrey here. `StoreUpgrades` shouldn't be part of the `setUpgradeHandler`. It should be directly passed from `app.go` and this info must be available without any calls to `KVStore`.

aaronc commented:
> > This could also be more self-contained if we added a store upgrades parameter to `SetHandler`:
> > ```go
> > upgradeKeeper.SetHandler("next-update", migrationFunc, storeUpgrades)
> > app.SetStoreLoader(upgradeKeeper.UpgradeStoreLoader())
> > ```
> 
> I agree with @ethanfrey here. `StoreUpgrades` shouldn't be part of the `setUpgradeHandler`. It should be directly passed from `app.go` and this info must be available without any calls to `KVStore`.

I'm not sure what the difference would be here. `SetUpgradeHandler` is called directly from `app.go` and this information would not be dependent on the `KVStore`. Wherever they get passed in, there needs to be a check based on the json file, not kvstore as to whether the store upgrades need to be applied. But it doesn't need to be in `SetUpgradeHandler` if that's not the right place. Do you have a proposed alternative that you could describe or point to @anilCSE ?

anilCSE commented:
> I'm not sure what the difference would be here. `SetUpgradeHandler` is called directly from `app.go` and this information would not be dependent on the `KVStore`. Wherever they get passed in, there needs to be a check based on the json file, not kvstore as to whether the store upgrades need to be applied. But it doesn't need to be in `SetUpgradeHandler` if that's not the right place. Do you have a proposed alternative that you could describe or point to @anilCSE ?

Ah, I missed to mention that the following suggestion works best. The current implementation follows the same. 

> So this could be as simple as something like:
> 
> ```go
> upgradeName, upgradeHeight := upgradeKeeper.ReadUpgradeInfoFromDisk()
> if upgradeName == "myUpgrade" {
>   // configure store loader that checks if version == upgradeHeight and applies store upgrades
>   app.SetStoreLoader(UpgradeStoreLoader(upgradeHeight, storeUpgrades)
> }
> ```

anilCSE commented:
The sample code-snippet to handle this could be,

```
// this configures a no-op upgrade handler for the "fooupgrade" upgrade
app.UpgradeKeeper.SetUpgradeHandler("fooupgrade", UpgradeHandlerFunc)

upgradeName, upgradeHeight := app.UpgradeKeeper.ReadUpgradeInfoFromDisk()
if upgradeName == "fooupgrade" && !app.UpgradeKeeper.IsSkipHeight(upgradeHeight) {
	storeUpgrades := store.StoreUpgrades{
		Renamed: []store.StoreRename{{
			OldKey: "foo",
			NewKey: "bar",
		}},
		Deleted: []string{},
	}

	// configure store loader that checks if version == upgradeHeight and applies store upgrades
	app.SetStoreLoader(upgrade.UpgradeStoreLoader(upgradeHeight, &storeUpgrades))
}
```

aaronc commented:
That solution makes sense @anilCSE. This is a pretty rare case so a bit of overhead in app.go is fine.

aaronc commented on x/upgrade/internal/keeper/keeper.go:
```diff
@@ -130,3 +139,65 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+// it assumes that the upgrade info is not available
+func (k Keeper) ReadUpgradeInfoFromDisk() (upgradeName string, upgradeHeight int64) {
+	var upgradeInfo store.UpgradeInfo
+	upgradeInfoPath := filepath.Join(k.homePath, UpgradeInfoFileName)
```
I think this should actually be stored under `data/upgrade-info.json`. I think that would be more consistent with how the Cosmos SDK currently works.

anilCSE replied:
Its updated now

anilCSE commented:
@alexanderbez can you check this PR

alexanderbez commented on simapp/app.go:
```diff
@@ -189,7 +192,7 @@ func NewSimApp(
 		app.subspaces[crisis.ModuleName], invCheckPeriod, app.SupplyKeeper, auth.FeeCollectorName,
 	)
-	app.UpgradeKeeper = upgrade.NewKeeper(skipUpgradeHeights, keys[upgrade.StoreKey], app.cdc)
+	app.UpgradeKeeper = upgrade.NewKeeper(skipUpgradeHeights, keys[upgrade.StoreKey], app.cdc, viper.GetString(flags.FlagHome))
```
The viper dep should not be used here. We need to figure out a way to provide the home value app constructor instead.

Something we need to address in another PR is really cleaning up the app constructor. We should use variadic option functions maybe.

anilCSE replied:
This might help: https://github.com/cosmos/cosmos-sdk/issues/5548

alexanderbez replied:
Yes, we should address that issue. But I would suggest doing it separately to keep this PR as lean as possible. For the time being, simply have the SimApp constructor take a new parameter or even introduce variadic options.

alexanderbez commented on store/types/store.go:
```diff
@@ -44,6 +44,13 @@ type StoreUpgrades struct {
 }
 
+// UpgradeInfo defines height and name of the upgrade
+// to ensure multistore upgrades happens only at matching height
```
```suggestion
// to ensure multistore upgrades happen only at matching height.
```

alexanderbez commented on x/upgrade/abci.go:
```diff
@@ -38,6 +38,14 @@ func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
 			ctx.Logger().Error(upgradeMsg)
+
+			// Write upgrade info to disk
+			// UpgradeStoreLoader uses this info to perform or skip store migrations
```
```
// Write the upgrade info to disk. The UpgradeStoreLoader uses this info to perform or skip
// store migrations.

alexanderbez commented on x/upgrade/abci_test.go:
```diff
@@ -393,3 +397,26 @@ func TestUpgradeWithoutSkip(t *testing.T) {
+	upgradeInfoFilePath, err := s.keeper.GetUpgradeInfoPath()
+	require.Nil(t, err)
+
+	data, err := ioutil.ReadFile(upgradeInfoFilePath)
```
Is this file removed after the test is complete? If not, can you ensure it's deleted?

alexanderbez commented on x/upgrade/internal/keeper/keeper.go:
```diff
@@ -130,3 +140,66 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
 	return k.skipUpgradeHeights[height]
 }
+
+// DumpUpgradeInfoToDisk adds plan height to upgrade-info.json
```
```suggestion
// DumpUpgradeInfoToDisk writes upgrade information to UpgradeInfoFileName.
```

alexanderbez commented on x/upgrade/internal/keeper/keeper.go:
```diff
@@ -130,3 +140,66 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+		Height: height,
+	}
+	info, err := json.Marshal(upgradeInfo)
+
```
```suggestion
```

alexanderbez commented on x/upgrade/internal/keeper/keeper.go:
```diff
@@ -130,3 +140,66 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+func (k Keeper) GetUpgradeInfoPath() (string, error) {
+	upgradeInfoFileDir := path.Join(k.GetHomeDir(), "data")
+
+	if _, err := os.Stat(upgradeInfoFileDir); os.IsNotExist(err) {
```
```go
import 	tmos "github.com/tendermint/tendermint/libs/os"
```

Use `tmos.EnsureDir` here instead of duplicating code 👍

alexanderbez commented on x/upgrade/internal/keeper/keeper.go:
```diff
@@ -130,3 +140,66 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+}
+
+// GetHomeDir returns the height at which the given upgrade was executed
+func (k Keeper) GetHomeDir() string {
```
Does this need to be exposed/used anywhere outside the keeper?

anilCSE replied:
Not really. I will change that

alexanderbez commented on x/upgrade/internal/keeper/keeper.go:
```diff
@@ -130,3 +140,66 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
+
+	var upgradeInfo store.UpgradeInfo
+	json.Unmarshal(data, &upgradeInfo)
+	return upgradeInfo.Name, upgradeInfo.Height
```
You're returning all the values from the `UpgradeInfo`  type. You might as well just return the entire `UpgradeInfo` struct which will cleanup the method signature 👍

anilCSE replied:
Updated the required changes

alexanderbez commented on x/upgrade/spec/01_concepts.md:
```diff
@@ -56,6 +56,17 @@ During each `EndBlock` execution, the `x/upgrade` module checks if there exists
 `Handler` is executed. If the `Plan` is expected to execute but no `Handler` is registered
 or if the binary was upgraded too early, the node will gracefully panic and exit.
 
+## StoreLoader
```
```suggestion
## StoreLoader

```

alexanderbez commented on x/upgrade/spec/01_concepts.md:
```diff
@@ -56,6 +56,17 @@ During each `EndBlock` execution, the `x/upgrade` module checks if there exists
 or if the binary was upgraded too early, the node will gracefully panic and exit.
 
+## StoreLoader
+The `x/upgrade` module also facilitates store migrations as part of the upgrade. The
```
I think we certainly need way more detail and explanation here. This is a core component of the upgrade procedure. I'd elaborate and explain concepts and procedures such as when and how a file is written to disk. What the file contains and why. How the file is used to migrate state....etc.

anilCSE commented:
> ACK -- just need to resolve the merge conflict ;)

Updated
